### PR TITLE
Dio driver task

### PR DESCRIPTION
**Pull request by M-Wagdy**

### CODE REVIEW

**Review by MohSamyN — APPROVED** [review]
Excellent design.

File: DIO_Driver/DIO_Driver/Mcal/DIO/DIO.c [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+* @param [in]  port     -  Port number.
+* @param [in]  pin      -  Pin number.
+* @param [out] data     -  pointer to the variable to save the pin value in it.
+*
+* @return function error state.
+*/
+DIO_ERROR_state_t DIO_ReadPin(uint8_t port, uint8_t pin, uint8_t * data)
+{
+   /* pointers to carry the required port registers address. */
+   volatile uint8_t * prt_port_stat = NULL_PTR;
```
Comment (AsaadKam) [inline DIO_Driver/DIO_Driver/Mcal/DIO/DIO.c:192]: -local variable defination:
You are not following Sprints_CodeGuideLine

File: DIO_Driver/DIO_Driver/Mcal/DIO/DIO.c [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+DIO_ERROR_state_t DIO_TogglePin(uint8_t port, uint8_t pin)
+{
+   /* check if invalid pin is given. */
+   if(PIN_7 < pin)
+   {
+      return E_DIO_INVALID_PIN;
+   }
+
+   /* pointers to carry the required port registers address. */
+   volatile uint8_t * prt_port_data = NULL_PTR;
```
Comment (AsaadKam) [inline DIO_Driver/DIO_Driver/Mcal/DIO/DIO.c:259]: -local variable defination:
You are not following Sprints_CodeGuideLine

File: DIO_Driver/DIO_Driver/Mcal/DIO/DIO_retval.h [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+/*- PRIMITIVE TYPES ----------------------------------------*/
+typedef uint8_t DIO_ERROR_state_t;
+
+/*- ERRORS DEFINITIONS--------------------------------------*/
+
+/**
+* @brief: This return value indicate that the function
+* ran successfully and it returns no errors.
+*/
+#define           E_DIO_SUCCESS              (DIO_ERROR_state_t)( 0 )
```
Comment (AsaadKam) [inline DIO_Driver/DIO_Driver/Mcal/DIO/DIO_retval.h:24]: -Error Handling should be like this formula in the image below



[Uploading Sprints_CodeGuideLine.pdf…]()

**Review by AsaadKam — COMMENTED** [review]
The code overall is very good , but you should take care from theses small mistakes

**Pull request merged**